snowflake_connector.py

```diff
@@ -0,0 +1,445 @@
+
+    def _handle_test_connectivity(self, param):
+        self.save_progress(TEST_CONNECTIVITY_PROGRESS_MSG)
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        try:
+            self._connection = self._handle_create_connection()
+            cursor = self._connection.cursor()
+
+            cursor.execute(SNOWFLAKE_VERSION_QUERY)
+            if cursor:
+                self.save_progress(TEST_CONNECTIVITY_SUCCESS_MSG)
+                return action_result.set_status(phantom.APP_SUCCESS)
+        except Exception as e:
+            self.save_progress(self._get_error_msg_from_exception(e))
+            return action_result.set_status(phantom.APP_ERROR, TEST_CONNECTIVITY_ERROR_MSG)
+
+    def _handle_run_query(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        query = param['query']
+        role = param.get('role')
+        warehouse = param.get('warehouse')
+        database = param.get('database')
+        schema = param.get('schema')
+
+        try:
+            self._connection = self._handle_create_connection(role, warehouse, database, schema)
+            cursor = self._connection.cursor(snowflake.connector.DictCursor)
+            cursor.execute(query)
+            returned_rows = cursor.fetchmany(DEFAULT_NUM_ROWS_TO_FETCH)
+
+            for row in returned_rows:
+                action_result.add_data(self._cleanup_row_values(row))
+
+            self.debug_print("returned_rows: {}".format(returned_rows))
+
+            while len(returned_rows) > 0:
+                returned_rows = cursor.fetchmany(DEFAULT_NUM_ROWS_TO_FETCH)
+                for row in returned_rows:
+                    action_result.add_data(self._cleanup_row_values(row))
+        except Exception as e:
+            error_msg = self._get_error_msg_from_exception(e)
+            self.save_progress("Error: {}".format(error_msg))
+            return action_result.set_status(phantom.APP_ERROR, '{0}: {1}'.format(SQL_QUERY_ERROR_MSG, error_msg))
+        finally:
+            if self._connection:
+                cursor.close()
+                self._connection.close()
+
+        summary = action_result.update_summary({})
+
+        if cursor.rowcount > 0:
+            summary[SNOWFLAKE_TOTAL_ROWS_JSON] = cursor.rowcount
+        else:
+            summary[SNOWFLAKE_TOTAL_ROWS_JSON] = 0
+
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def _handle_disable_user(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        database = SNOWFLAKE_DATABASE
+        username = param['username']
+        role = param.get('role')
+
+        try:
+            self._connection = self._handle_create_connection(database=database, role=role)
```

Comment:
kind of a general question since `role` is optional for all of these actions: are there cases where you don't need to specify a role to establish a connection? does it just know to default to some role or something?

Reply:
Yes, there are two cases:it is possible that there is a DEFAULT_ROLE assigned to a user and/or the object being queried is owned by the user. In either of those cases,

Reply:
Yes, there are two cases: 1) there can be a DEFAULT_ROLE assigned to a user and/or 2) the object being queried is owned by that user. In either of these cases, the role parameter is not needed. Typically you could also execute a SQL statement in the same session like 'use role accountadmin;' to change roles, but since we are only allowing one SQL statement per session it is being passed when the connection is made.